primitives/runtime/src/traits.rs

```diff
@@ -530,18 +530,68 @@ morph_types! {
 	} where L::Type: Ord, M: TryMorph<L::Type, Outcome = L::Type>;
 }
 
-/// Extensible conversion trait. Generic over both source and destination types.
+/// Infallible conversion trait. Generic over both source and destination types.
 pub trait Convert<A, B> {
 	/// Make conversion.
 	fn convert(a: A) -> B;
 }
 
+/// Fallible conversion trait returning a Result. Generic over both source and destination types.
+///
+/// In case an `Error` is returned, it must alwats be the operand which is returned.
+pub trait TryConvert<A, B> {
+	/// Make conversion.
+	fn try_convert(a: A) -> Result<B, A>;
+}
+
+/// Reversing infallible conversion trait. Generic over both source and destination types.
+///
+/// This specifically reverses the conversion.
+pub trait ConvertBack<A, B>: Convert<A, B> {
+	/// Make conversion back.
+	fn convert_back(b: B) -> A;
+}
+
+/// Fallible conversion trait returning a Result. Generic over both source and destination types.
+///
+/// In case an `Error` is returned, it must alwats be the operand which is returned.
+pub trait TryConvertBack<A, B>: TryConvert<A, B> {
+	/// Make conversion.
+	fn try_convert_back(b: B) -> Result<A, B>;
+}
+
 impl<A, B: Default> Convert<A, B> for () {
 	fn convert(_: A) -> B {
 		Default::default()
 	}
 }
 
+#[impl_trait_for_tuples::impl_for_tuples(30)]
+impl<A, B> TryConvert<A, B> for Tuple {
```

Comment:
I have a hard time imagining how usage of this looks like.  
Could you maybe add a minimal test of `(u8, u8) -> (u16, u16)` or similar?

Reply:
The idea is `TryConvert<u8, u16> for (TypeX, TypeY)`. This isn't about making `A` and `B` tuples (which is also valid, even without this impl).